File: cpp/02_medium/src/ScoreManager.cpp
#include "../include/ScoreManager.h"
#include <fstream>
#include <iostream>

ScoreManager::ScoreManager() : currentScore(0), highScore(0) {
    loadHighScore();
}

void ScoreManager::loadHighScore() {
    std::ifstream file("highscore.txt");
    if (file.is_open()) {
        file >> highScore;
        file.close();
    }
}

void ScoreManager::saveHighScore() {
    std::ofstream file("highscore.txt");
    if (file.is_open()) {
        file << highScore;
        file.close();
    }
}

void ScoreManager::addScore(int points) {
    currentScore += points;
    if (currentScore > highScore) {
        highScore = currentScore;
        saveHighScore();
    }
}

void ScoreManager::resetScore() {
    currentScore = 0;
}

File: csharp/02_medium/ScoreManager.cs
using System;
using System.IO;
using System.Text.Json;

namespace SnakeGameWinForms
{
    public class ScoreManager
    {
        private const string ScoresFile = "scores.json";
        private int currentScore;
        private int highScore;

        public int CurrentScore => currentScore;
        public int HighScore => highScore;

        public ScoreManager()
        {
            LoadScores();
        }

        public void AddScore(int points)
        {
            currentScore += points;
            if (currentScore > highScore)
            {
                highScore = currentScore;
                SaveScores();
            }
        }

        public void ResetScore()
        {
            currentScore = 0;
        }

        private void LoadScores()
        {
            try
            {
                if (File.Exists(ScoresFile))
                {
                    string json = File.ReadAllText(ScoresFile);
                    var data = JsonSerializer.Deserialize<ScoreData>(json);
                    if (data != null)
                    {
                        highScore = data.HighScore;
                    }
                }
            }
            catch
            {
                highScore = 0;
            }
        }

        private void SaveScores()
        {
            try
            {
                var data = new ScoreData { HighScore = highScore };
                string json = JsonSerializer.Serialize(data);
                File.WriteAllText(ScoresFile, json);
            }
            catch { }
        }

        private class ScoreData
        {
            public int HighScore { get; set; }
        }
    }
}
